Library/Homebrew/utils/spdx.rb

```diff
@@ -5,19 +5,111 @@
   end
 
   def download_latest_license_data!(to: JSON_PATH)
-    data_url = "https://raw.githubusercontent.com/spdx/license-list-data/#{latest_tag}/json/licenses.json"
-    curl_download(data_url, to: to, partial: false)
+    data_url = "https://raw.githubusercontent.com/spdx/license-list-data/#{latest_tag}/json/"
+    curl_download(data_url + "licenses.json", to: to/"spdx_licenses.json", partial: false)
+    curl_download(data_url + "exceptions.json", to: to/"spdx_exceptions.json", partial: false)
+  end
+
+  def parse_license_expression(license_expression)
+    licenses = []
+    exceptions = []
+
+    case license_expression
+    when String, Symbol
+      licenses.push license_expression
+    when Hash
+      key = license_expression.keys.first
+      if [:any_of, :all_of].include? key
+        sub_license, sub_exception = parse_license_expression license_expression[key]
+        licenses += sub_license
+        exceptions += sub_exception
+      else
+        licenses.push key
+        exceptions.push license_expression[key][:with]
+      end
+    when Array
+      license_expression.each do |license|
+        sub_license, sub_exception = parse_license_expression license
+        licenses += sub_license
+        exceptions += sub_exception
+      end
+    end
+
+    [licenses, exceptions]
+  end
+
+  def valid_license?(license)
+    return true if license == :public_domain
+
+    license = license.delete_suffix "+"
+    license_data["licenses"].any? { |spdx_license| spdx_license["licenseId"] == license }
+  end
+
+  def deprecated_license?(license)
+    return false if license == :public_domain
+    return true unless valid_license?(license)
+
+    license_data["licenses"].none? do |spdx_license|
+      spdx_license["licenseId"] == license && !spdx_license["isDeprecatedLicenseId"]
+    end
+  end
+
+  def valid_license_exception?(exception)
+    exception_data["exceptions"].any? do |spdx_exception|
+      spdx_exception["licenseExceptionId"] == exception && !spdx_exception["isDeprecatedLicenseId"]
+    end
+  end
+
+  def license_expression_to_string(license_expression, bracket: false)
+    out = ""
+
+    case license_expression
+    when String
+      out += license_expression
+    when :public_domain
+      out += "Public Domain"
+    when Hash
+      out += "(" if bracket
+      key = license_expression.keys.first
+      if [:any_of, :all_of].include? key
+        licenses = license_expression[key].map { |license| license_expression_to_string license, bracket: true }
+        out += licenses.join key == :any_of ? " OR " : " AND "
+      else
+        out += "#{key} WITH #{license_expression[key][:with]}"
+      end
+      out += ")" if bracket
+    end
+
+    out
+  end
+
+  def license_version_info(license)
+    return [license] if license == :public_domain
+
+    match = license.match(/-([0-9.]+)(?:-.*?)??(\+|-only|-or-later)?$/)
+    return [license] unless match.present?
+
+    license_name = license.split(match[0]).first
```

Comment:
Use named groups for these to improve readability: https://ruby-doc.org/core-2.7.0/Regexp.html